processor/stream/stream_processor.go

```diff
@@ -229,6 +239,25 @@ func (s *StreamProcessor) HandleStream(ctx context.Context, meta map[string]inte
 	}
 
 	for {
+
+		//Throttle read in case rate limit is hit, by using
+		//blocking wait until rate limiter allows to read next batchsize event.
+		//Consider burst and timeout indicating how long to wait for rate limiter.
+		//Return early if rate limiter returns error.
+		limiter, limiterOk := ctx.Value(RateLimiterKey).(*rate.Limiter)
+		if limiterOk && limiter != nil {
+			ctxT, cancel := context.WithTimeout(ctx, time.Second)
+			defer cancel()
+			err = limiter.WaitN(ctxT, batchSize)
```

Comment:
this decreases the rate limit "bucket" by `batchSize` even if we read fewer items for every batch. 
We could consider doing something like:

```
readsAllowed := 0
for {
    if readAllowed < batchSize {
        missing := batchSize - readAllowed
        limiter.WaitN(ctxT, missing)
        readsAllowed += missing
    }
    n, transformables, done := s.readBatch(batchSize, jsonReader, res)
    readsAllowed = readsAllowed - n
    [...]
}
```

I also think you should move: 
`limiter, limiterOk := ctx.Value(RateLimiterKey).(*rate.Limiter)`
out of the loop. No reason to get it from the context every time. 

finally, `defer` defers the execution of `cancel()` to when the function returns, not until the loop does another iteration. We'll need to call `cancel()` manually here.

Reply:
could you instead move the rate limiting into `readBatch`, before reach call to StreamReader.Read? or better yet, wrap StreamReader such that its Reads are rate-limited? e.g.

```diff
diff --git a/processor/stream/stream_processor.go b/processor/stream/stream_processor.go
index 6182d0cc..d284f6a2 100644
--- a/processor/stream/stream_processor.go
+++ b/processor/stream/stream_processor.go
@@ -23,6 +23,7 @@ import (
        "io"
 
        "github.com/santhosh-tekuri/jsonschema"
+       "golang.org/x/time/rate"
 
        "github.com/elastic/apm-server/decoder"
        er "github.com/elastic/apm-server/model/error"
@@ -74,6 +75,19 @@ func (s *srErrorWrapper) Read() (map[string]interface{}, error) {
        return v, err
 }
 
+type rateLimitStreamReader struct {
+       StreamReader
+       ctx context.Context
+       rl  *rate.Limiter
+}
+
+func (r *rateLimitStreamReader) Read() (map[string]interface{}, error) {
+       if err := r.Wait(r.ctx); err != nil {
+               return nil, err
+       }
+       return r.StreamReader.Read()
+}
+
 type StreamProcessor struct {
        Tconfig transform.Config
 }
@@ -222,6 +236,16 @@ func (s *StreamProcessor) HandleStream(ctx context.Context, meta map[string]inte
                return res
        }
 
+       if rl, _ := ctx.Value(rateLimitKey{}); rl != nil {
+               // A rate-limiter has been provided; wrap the stream
+               // reader so that reads are rate-limited.
+               jsonReader = &rateLimitStreamReader{
+                       StreamReader: jsonReader,
+                       ctx:          ctx,
+                       rl:           rl,
+               }
+       }
+
        tctx := &transform.Context{
                RequestTime: utility.RequestTime(ctx),
                Config:      s.Tconfig,
```

Reply:
By the way, you might also want to take a look at https://github.com/juju/ratelimit as an alternative rate limiter. It's a little more flexible, and I think probably more efficient.

e.g. to rate-limit, but wait for no more than 1 second, with juju/ratelimit you would do this:

```go
var r *ratelimit.Bucket = ...
if !r.WaitMaxDuration(batchSize, time.Second) {
    res.Add(&Error{Type: RateLimitErrType, Message: "rate limit exceeded"})
    return res
}
```

Reply:
@axw in my first draft I did call `Wait` on every iteration. I was actually thinking about just wrapping it around the reader. However, we would have to check for every single event, which decreases performance. After discussing with @roncohen he suggested to check for `batchSize` events at once and accept to 'waste' some requests for the last batch that is read. 

@roncohen I don't agree with your suggestion [here](https://github.com/elastic/apm-server/pull/1367/files/2f0de7b8e53bd44cfe9fa09d2b04824109038483#r217454248). 
The only reason for not reading `batchsize` events in a batch would be if there was an early-return-error or if it is the last batch that had less than batchsize entries. In both cases, we would return without an additional iteration. In any other case `readBatch` should have read `batchsize` entries. So the case that we waste `batchsize-n` hits always only applies to the last batch. This would not be solved by your suggestion.

Reply:
ah, true 👍

Reply:
@axw update on `Wait` vs. `Reserve`: Using `Reserve` leads to *always* reserve a token if the amount of checked tokens is < limit.burst, which is always the case when checking per event. See [`ReserveN` always calls `reserveN` with `InfDuration`](https://github.com/golang/time/blob/master/rate/rate.go#L211), and therefore [`reserveN` always updating state](https://github.com/golang/time/blob/master/rate/rate.go#L330).
There is no way of adding a timeout to `Reserve` or `ReserveN`. Using `Wait` would only update, if the `deadline` is met. This means I cannot use `Reserve` for throttling the read. As soon as one request hits the limit, within a second all following requests would fail too, and it cannot relax anymore, so all requests would return with a rate limit hit error.